crud/src/main/java/com/redhat/lightblue/assoc/RewriteQuery.java

```diff
@@ -0,0 +1,496 @@
+                if(q.getOp()==NaryLogicalOperator._and) {
+                    if(nestedq instanceof TruePH) {
+                        itr.remove();
+                    } else if(nestedq instanceof FalsePH) {
+                        return new FalsePH();
+                    } 
+                } else {
+                    if(nestedq instanceof TruePH) {
+                        return new TruePH();
+                    } else if(nestedq instanceof FalsePH) {
+                        itr.remove();
+                    } 
+                }
+            }
+            if(itr.isCopied()) {
+                List<QueryExpression> newList=itr.getCopiedList();
+                if(newList.size()==0)
+                	return new TruePH();
+                else if(newList.size()==1)
+                    return newList.get(0);
+                else
+                    return new NaryLogicalExpression(q.getOp(),newList);
+            } else {
+                return nq;
+            }
+        }
+
+        /**
+         * Several possibilities exist when rewriting an array query.
+         *
+         * The array field itself may be pointing to a different entity
+         * than the currentEntity. This includes the case where the array
+         * is the reference field itself. If this is the case, this is no
+         * longer an array elem match expression. The nested query is
+         * moved outside and rewritten.
+         *
+         * The array field itself may be pointing to an array in
+         * currentEntity. In this case, the array field is rewritten as
+         * the local field, and the nested query is rewritten recursively.
+         *
+         * If the nested query contains a placeholder, the query is
+         * replaced with that. Otherwise, the query remains as is
+         */
+        @Override
+        protected QueryExpression itrArrayMatchExpression(ArrayMatchExpression q, Path context) {
+            QueryFieldInfo qfi=findFieldInfo(q.getArray(),q);
+            if(qfi.getFieldEntity()==currentEntity) {
+                if(qfi.getFieldMd() instanceof ResolvedReferenceField) {
+                    // The array is pointing to a reference field
+                    // This could be a simple reference, like {array:ref, elemMatch: Q }
+                    // In that case, the return value is Q
+                    // Example: in: { array: ref, elemMatch:{ field:_id, op:$eq, rfield: $parent.ref_id} }
+                    //         out: { field: ref_id, op: $eq, rvalue: <_id value> }
+                    //
+                    // This could also be a nested array reference, like {array: x.*.ref, elemMatch: Q }
+                    // Example: in: { array: x.*.obj.ref, elemMatch: { field: _id, op:$eq, rfield: $parent.ref_id } }
+                    //         out: { array: x, elemMatch: { field: obj.ref_id, op: $eq, rvalue: < _id value> } }
+
+                    // Find the closes array containing the reference
```

Comment:
fixed.

On Tue, Mar 29, 2016 at 10:42 AM, Brandon Vulaj <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/RewriteQuery.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57756438>
> :
>
> > +        @Override
> > +        protected QueryExpression itrArrayMatchExpression(ArrayMatchExpression q, Path context) {
> > +            QueryFieldInfo qfi=findFieldInfo(q.getArray(),q);
> > +            if(qfi.getFieldEntity()==currentEntity) {
> > +                if(qfi.getFieldMd() instanceof ResolvedReferenceField) {
> > +                    // The array is pointing to a reference field
> > +                    // This could be a simple reference, like {array:ref, elemMatch: Q }
> > +                    // In that case, the return value is Q
> > +                    // Example: in: { array: ref, elemMatch:{ field:_id, op:$eq, rfield: $parent.ref_id} }
> > +                    //         out: { field: ref_id, op: $eq, rvalue: <_id value> }
> > +                    //
> > +                    // This could also be a nested array reference, like {array: x.*.ref, elemMatch: Q }
> > +                    // Example: in: { array: x.*.obj.ref, elemMatch: { field: _id, op:$eq, rfield: $parent.ref_id } }
> > +                    //         out: { array: x, elemMatch: { field: obj.ref_id, op: $eq, rvalue: < _id value> } }
> > +
> > +                    // Find the closes array containing the reference
>
> Nitpick: spelling
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/4bcf56affb2ffd020fa4c380a7af065c40010abf#r57756438>
>